ambersim/utils/io_utils.py

```diff
@@ -47,57 +48,80 @@ def _modify_robot_float_base(filepath: Union[str, Path]) -> mj.MjModel:
+    # only add free joint if the first body after worldbody has no freejoints
     assert robot.worldbody is not None
-    if len(robot.worldbody.body[0].joint) == 0:
+    if robot.worldbody.body[0].freejoint is None:
+        # creates a new robot object and attaches the current robot to it freely
         arena = mjcf.RootElement(model=robot.model)
         attachment_frame = arena.attach(robot)
         attachment_frame.add("freejoint", name="freejoint")
         robot = arena
-    model = mj.MjModel.from_xml_string(robot.to_xml_string())
+
+        # ensuring that the base link has inertial attributes
+        assert robot.worldbody is not None  # pyright typechecking
+        if robot.worldbody.body[0].inertial is None:
+            robot.worldbody.body[0].add("inertial", mass=1e-6, diaginertia=1e-6 * np.ones(3), pos=np.zeros(3))
+
+    # extracts the mujoco model from the dm_mujoco Physics object
+    physics = mjcf.Physics.from_mjcf_model(robot)
+    assert physics is not None  # pyright typechecking
+    model = physics.model.ptr
     return model
 
 
-def load_mjx_model_from_file(filepath: Union[str, Path], force_float: bool = False) -> Tuple[mjx.Model, mjx.Data]:
-    """Loads a mjx model/data pair from a filepath.
+def load_mj_model_from_file(filepath: Union[str, Path], force_float: bool = False) -> mj.MjModel:
+    """Loads a mujoco model from a filepath.
 
     Args:
         filepath: A path to a URDF or MJCF file. This can be global, local, or with respect to the repository root.
         force_float: Whether to forcibly float the base of the robot.
 
     Returns:
-        mjx_model: A mjx Model.
-        mjx_data: A mjx Data struct.
+        mj_model: A mujoco model.
     """
     filepath = _check_filepath(filepath)
 
     # loading the model and data. check whether freejoint is added forcibly
     if force_float:
         # check that the file is an XML. if not, save as xml temporarily
         if str(filepath).split(".")[-1] != "xml":
-            save_model_xml(filepath, output_name="_temp_xml_model")
-            _model = _modify_robot_float_base("_temp_xml_model.xml")
-            Path.unlink("_temp_xml_model.xml")
+            output_name = "/".join(str(filepath).split("/")[:-1]) + "/_temp_xml_model.xml"
+            save_model_xml(filepath, output_name=output_name)
+            mj_model = _modify_robot_float_base(output_name)
+            Path.unlink(output_name)
         else:
-            _model = _modify_robot_float_base(filepath)
+            mj_model = _modify_robot_float_base(filepath)
     else:
-        _model = mj.MjModel.from_xml_path(filepath)
+        mj_model = mj.MjModel.from_xml_path(filepath)
+    return mj_model
+
 
+def mj_to_mjx_model(mj_model: mj.MjModel) -> Tuple[mjx.Model, mjx.Data]:
+    """Converts a mujoco model to an mjx (model, data) pair."""
     try:
-        mjx_model = mjx.device_put(_model)
-        mjx_data = mjx.make_data(_model)
+        mjx_model = mjx.device_put(mj_model)
```

Comment:
When I run this I get warnings about MJX not supporting the Newton solver. It looks from the [website](https://mujoco.readthedocs.io/en/latest/mjx.html) like MJX does support Newton now, but I wasn't able to get that version from pip. Do we want to force MJX to use the CG solver, e.g., `mj_model.opt.solver = mj.mjtSolver.mjSOL_CG`?

Reply:
It should by default force it to CG solver right now on MJX side. But they also have already implemented on Newton solver it's just not on their stable release yet

Reply:
I opened an issue last night about allowing `mujoco` install from source instead of vs. `pip` because the nightly release has a lot of new upgrades over `3.0.0` including support for the Newton solver. I basically never want to force the CG solver once this becomes the official `pypi` release, so maybe this can get resolved in a separate PR?

In the meantime, re-installing `mujoco` from github on your local env directly should kill the warning.

Reply:
Right, just wondering if we should make that conversion explicit so we don't get warnings when we run `pytest`. I want those pretty green checkmarks!

Reply:
I can expose the option for the solver in the API and add a TODO to come back and make Newton the default once the other PR is done + add this to that issue

Reply:
Exposed:
```
def load_mj_model_from_file(
    filepath: Union[str, Path],
    force_float: bool = False,
    solver: Union[str, mj.mjtSolver] = mj.mjtSolver.mjSOL_CG,
    iterations: Optional[int] = None,
    ls_iterations: Optional[int] = None,
) -> mj.MjModel:
    """Loads a mujoco model from a filepath.
    Args:
        filepath: A path to a URDF or MJCF file. This can be global, local, or with respect to the repository root.
        force_float: Whether to forcibly float the base of the robot.
        solver: The type of solver to use. For now, mjx only supports CG.
        iterations: The number of solver iterations to use. If unspecified, mujoco uses 100.
        ls_iterations: The number of line search iterations to use. If unspecified, mujoco uses 50.
    Returns:
        mj_model: A mujoco model.
    """
    # TODO(ahl): once we allow installing mujoco from source, update this to allow the Newton solver + update default
    if isinstance(solver, str):
        if solver.lower() == "cg":
            solver = mj.mjtSolver.SOL_CG
        else:
            raise ValueError("Solver must be one of: ['cg']!")
    elif isinstance(solver, mj.mjtSolver):
        assert solver in [mj.mjtSolver.mjSOL_CG]
```